File: apps/shared/src/voice-contract.ts
export const TRANSCRIPTION_SCHEMA = "fabric.transcription" as const;
export const TRANSCRIPTION_VERSION = 1 as const;
export const PHONE_AUDIO_CONTRACT = "fabric.phone_audio" as const;
export const PHONE_AUDIO_VERSION = 1 as const;

export const TRANSCRIPTION_STATUSES = [
  "completed",
  "no_speech",
  "cancelled",
  "failed",
] as const;
export const PHONE_AUDIO_MODES = [
  "dictate",
  "voice_note",
  "ask_fabric",
  "chat",
] as const;

export type TranscriptionStatus = (typeof TRANSCRIPTION_STATUSES)[number];
export type PhoneAudioMode = (typeof PHONE_AUDIO_MODES)[number];

export interface TranscriptionSegmentV1 {
  start_ms: number;
  end_ms: number;
  text: string;
}

export interface TranscriptionErrorV1 {
  code: string;
  message: string;
  retryable: boolean;
}

export interface TranscriptionResultV1 {
  schema: typeof TRANSCRIPTION_SCHEMA;
  version: typeof TRANSCRIPTION_VERSION;
  request_id: string;
  status: TranscriptionStatus;
  text: string;
  provider?: string;
  language?: string;
  duration_ms?: number;
  processing_ms?: number;
  model?: string;
  segments: readonly TranscriptionSegmentV1[];
  warnings: readonly string[];
  error?: TranscriptionErrorV1;
}

export interface PhoneAudioEnvelopeV1 {
  contract: typeof PHONE_AUDIO_CONTRACT;
  version: typeof PHONE_AUDIO_VERSION;
  capture_id: string;
  mode: PhoneAudioMode;
  mime_type: string;
  duration_ms: number;
  result: TranscriptionResultV1;
}

export type VoiceContractParseResult<T> =
  | { kind: "verified"; value: T }
  | { kind: "incompatible"; contract: string; version: number }
  | { kind: "invalid"; message: string };

const MAX_AUDIO_MS = 3_600_000;
const MAX_TEXT_CHARS = 1_000_000;

class VoiceDecodeError extends Error {}
class VoiceIncompatibleError extends Error {
  readonly contract: string;
  readonly version: number;

  constructor(contract: string, version: number) {
    super(`${contract} version ${version} is incompatible.`);
    this.contract = contract;
    this.version = version;
  }
}

function fail(message: string): never {
  throw new VoiceDecodeError(message);
}

function record(value: unknown, path: string): Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    return fail(`${path} must be an object.`);
  }
  return value as Record<string, unknown>;
}

function required(
  raw: Record<string, unknown>,
  key: string,
  path: string,
): unknown {
  if (!Object.prototype.hasOwnProperty.call(raw, key)) {
    return fail(`${path}.${key} is required.`);
  }
  return raw[key];
}

function text(
  value: unknown,
  path: string,
  maximum: number,
  allowEmpty = false,
): string {
  if (typeof value !== "string") return fail(`${path} must be a string.`);
  if (!allowEmpty && !value.trim()) return fail(`${path} must not be empty.`);
  if (Array.from(value).length > maximum) return fail(`${path} is too long.`);
  return value;
}

function optionalText(
  raw: Record<string, unknown>,
  key: string,
  path: string,
  maximum: number,
): string | undefined {
  return raw[key] === undefined
    ? undefined
    : text(raw[key], `${path}.${key}`, maximum);
}

function integer(value: unknown, path: string, maximum = MAX_AUDIO_MS): number {
  if (
    !Number.isSafeInteger(value) ||
    (value as number) < 0 ||
    (value as number) > maximum
  ) {
    return fail(`${path} must be an integer between 0 and ${maximum}.`);
  }
  return value as number;
}

function parseTranscription(value: unknown): TranscriptionResultV1 {
  const raw = record(value, "transcription");
  if (required(raw, "schema", "transcription") !== TRANSCRIPTION_SCHEMA) {
    return fail("transcription.schema is unsupported.");
  }
  const version = required(raw, "version", "transcription");
  if (!Number.isSafeInteger(version))
    return fail("transcription.version must be an integer.");
  if (version !== TRANSCRIPTION_VERSION) {
    throw new VoiceIncompatibleError(TRANSCRIPTION_SCHEMA, version as number);
  }
  const requestId = text(
    required(raw, "request_id", "transcription"),
    "transcription.request_id",
    128,
  );
  const statusValue = required(raw, "status", "transcription");
  if (!TRANSCRIPTION_STATUSES.includes(statusValue as TranscriptionStatus)) {
    return fail("transcription.status is invalid.");
  }
  const status = statusValue as TranscriptionStatus;
  const transcript = text(
    required(raw, "text", "transcription"),
    "transcription.text",
    MAX_TEXT_CHARS,
    true,
  );
  const duration =
    raw.duration_ms === undefined
      ? undefined
      : integer(raw.duration_ms, "transcription.duration_ms");
  const processing =
    raw.processing_ms === undefined
      ? undefined
      : integer(raw.processing_ms, "transcription.processing_ms");

  const segmentValues = raw.segments ?? [];
  if (!Array.isArray(segmentValues) || segmentValues.length > 10_000) {
    return fail("transcription.segments must be a bounded array.");
  }
  const segments = segmentValues.map((value, index) => {
    const path = `transcription.segments[${index}]`;
    const segment = record(value, path);
    const start = integer(
      required(segment, "start_ms", path),
      `${path}.start_ms`,
    );
    const end = integer(required(segment, "end_ms", path), `${path}.end_ms`);
    if (end < start) return fail(`${path}.end_ms must not precede start_ms.`);
    if (duration !== undefined && end > duration) {
      return fail(`${path}.end_ms exceeds transcription.duration_ms.`);
    }
    return {
      start_ms: start,
      end_ms: end,
      text: text(
        required(segment, "text", path),
        `${path}.text`,
        MAX_TEXT_CHARS,
        true,
      ),
    };
  });

  const warningValues = raw.warnings ?? [];
  if (!Array.isArray(warningValues) || warningValues.length > 64) {
    return fail("transcription.warnings must be a bounded array.");
  }
  const warnings = warningValues.map((value, index) =>
    text(value, `transcription.warnings[${index}]`, 1000, true),
  );

  let transcriptionError: TranscriptionErrorV1 | undefined;
  if (status === "failed") {
    if (transcript) return fail("failed transcription text must be empty.");
    const error = record(
      required(raw, "error", "transcription"),
      "transcription.error",
    );
    const retryable = required(error, "retryable", "transcription.error");
    if (typeof retryable !== "boolean") {
      return fail("transcription.error.retryable must be a boolean.");
    }
    transcriptionError = {
      code: text(
        required(error, "code", "transcription.error"),
        "transcription.error.code",
        128,
      ),
      message: text(
        required(error, "message", "transcription.error"),
        "transcription.error.message",
        4000,
      ),
      retryable,
    };
  } else if (raw.error !== undefined) {
    return fail("only failed transcriptions may contain error.");
  } else if ((status === "no_speech" || status === "cancelled") && transcript) {
    return fail(`${status} transcription text must be empty.`);
  }

  return {
    schema: TRANSCRIPTION_SCHEMA,
    version: TRANSCRIPTION_VERSION,
    request_id: requestId,
    status,
    text: transcript,
    provider: optionalText(raw, "provider", "transcription", 128),
    language: optionalText(raw, "language", "transcription", 64),
    duration_ms: duration,
    processing_ms: processing,
    model: optionalText(raw, "model", "transcription", 128),
    segments,
    warnings,
    error: transcriptionError,
  };
}

function parseResult<T>(
  parser: () => T,
  fallback: string,
): VoiceContractParseResult<T> {
  try {
    return { kind: "verified", value: parser() };
  } catch (error) {
    if (error instanceof VoiceIncompatibleError) {
      return {
        kind: "incompatible",
        contract: error.contract,
        version: error.version,
      };
    }
    return {
      kind: "invalid",
      message: error instanceof VoiceDecodeError ? error.message : fallback,
    };
  }
}

export function parseTranscriptionResult(
  value: unknown,
): VoiceContractParseResult<TranscriptionResultV1> {
  return parseResult(
    () => parseTranscription(value),
    "transcription is malformed.",
  );
}

export function parsePhoneAudio(
  value: unknown,
): VoiceContractParseResult<PhoneAudioEnvelopeV1> {
  return parseResult(() => {
    const raw = record(value, "phone_audio");
    if (required(raw, "contract", "phone_audio") !== PHONE_AUDIO_CONTRACT) {
      return fail("phone_audio.contract is unsupported.");
    }
    const version = required(raw, "version", "phone_audio");
    if (!Number.isSafeInteger(version))
      return fail("phone_audio.version must be an integer.");
    if (version !== PHONE_AUDIO_VERSION) {
      throw new VoiceIncompatibleError(PHONE_AUDIO_CONTRACT, version as number);
    }
    const modeValue = required(raw, "mode", "phone_audio");
    if (!PHONE_AUDIO_MODES.includes(modeValue as PhoneAudioMode)) {
      return fail("phone_audio.mode is invalid.");
    }
    const mimeType = text(
      required(raw, "mime_type", "phone_audio"),
      "phone_audio.mime_type",
      128,
    ).toLowerCase();
    if (!mimeType.startsWith("audio/") && mimeType !== "video/webm") {
      return fail("phone_audio.mime_type must describe audio.");
    }
    return {
      contract: PHONE_AUDIO_CONTRACT,
      version: PHONE_AUDIO_VERSION,
      capture_id: text(
        required(raw, "capture_id", "phone_audio"),
        "phone_audio.capture_id",
        128,
      ),
      mode: modeValue as PhoneAudioMode,
      mime_type: mimeType,
      duration_ms: integer(
        required(raw, "duration_ms", "phone_audio"),
        "phone_audio.duration_ms",
      ),
      result: parseTranscription(required(raw, "result", "phone_audio")),
    };
  }, "phone_audio is malformed.");
}

File: apps/mobile/android/app/src/main/kotlin/io/github/obliviousodin/fabric/mobile/core/VoiceContract.kt
package io.github.obliviousodin.fabric.mobile.core

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.intOrNull

const val FABRIC_TRANSCRIPTION_CONTRACT_VERSION = 1
const val FABRIC_PHONE_AUDIO_CONTRACT_VERSION = 1

data class TranscriptionSegmentV1(
    val startMs: Int,
    val endMs: Int,
    val text: String,
)

data class TranscriptionErrorV1(
    val code: String,
    val message: String,
    val retryable: Boolean,
)

data class TranscriptionResultV1(
    val requestId: String,
    val status: String,
    val text: String,
    val provider: String?,
    val language: String?,
    val durationMs: Int?,
    val processingMs: Int?,
    val model: String?,
    val segments: List<TranscriptionSegmentV1>,
    val warnings: List<String>,
    val error: TranscriptionErrorV1?,
)

data class PhoneAudioEnvelopeV1(
    val captureId: String,
    val mode: String,
    val mimeType: String,
    val durationMs: Int,
    val result: TranscriptionResultV1,
)

sealed interface VoiceContractParseResult<out T> {
    data class Verified<T>(val value: T) : VoiceContractParseResult<T>
    data class Incompatible(val contract: String, val version: Int) : VoiceContractParseResult<Nothing>
    data class Invalid(val message: String) : VoiceContractParseResult<Nothing>
}

private class VoiceContractDecodeException(message: String) : IllegalArgumentException(message)
private class VoiceContractIncompatibleException(
    val contract: String,
    val version: Int,
) : IllegalArgumentException("$contract version $version is incompatible.")

private const val MAX_AUDIO_MS = 3_600_000
private const val MAX_TEXT_CHARACTERS = 1_000_000
private val TRANSCRIPTION_STATUSES = setOf("completed", "no_speech", "cancelled", "failed")
private val PHONE_AUDIO_MODES = setOf("dictate", "voice_note", "ask_fabric", "chat")

private fun fail(message: String): Nothing = throw VoiceContractDecodeException(message)

private fun objectValue(value: JsonElement?, path: String): JsonObject =
    value as? JsonObject ?: fail("$path must be an object.")

private fun required(value: JsonObject, key: String, path: String): JsonElement =
    value[key] ?: fail("$path.$key is required.")

private fun stringValue(
    value: JsonElement?,
    path: String,
    maximum: Int,
    allowEmpty: Boolean = false,
): String {
    val primitive = value as? JsonPrimitive ?: fail("$path must be a string.")
    if (!primitive.isString) fail("$path must be a string.")
    val result = primitive.content
    if (!allowEmpty && result.isBlank()) fail("$path must not be empty.")
    if (result.length > maximum) fail("$path is too long.")
    return result
}

private fun optionalString(value: JsonObject, key: String, path: String, maximum: Int): String? =
    value[key]?.let { stringValue(it, "$path.$key", maximum) }

private fun integerValue(value: JsonElement?, path: String): Int {
    val result = (value as? JsonPrimitive)?.intOrNull ?: fail("$path must be an integer.")
    if (result !in 0..MAX_AUDIO_MS) fail("$path is outside the supported range.")
    return result
}

private fun parseTranscription(value: JsonElement): TranscriptionResultV1 {
    val raw = objectValue(value, "transcription")
    if (stringValue(required(raw, "schema", "transcription"), "transcription.schema", 128) != "fabric.transcription") {
        fail("transcription.schema is unsupported.")
    }
    val version = (required(raw, "version", "transcription") as? JsonPrimitive)?.intOrNull
        ?: fail("transcription.version must be an integer.")
    if (version != FABRIC_TRANSCRIPTION_CONTRACT_VERSION) {
        throw VoiceContractIncompatibleException("fabric.transcription", version)
    }
    val requestId = stringValue(required(raw, "request_id", "transcription"), "transcription.request_id", 128)
    val status = stringValue(required(raw, "status", "transcription"), "transcription.status", 32)
    if (status !in TRANSCRIPTION_STATUSES) fail("transcription.status is invalid.")
    val text = stringValue(
        required(raw, "text", "transcription"),
        "transcription.text",
        MAX_TEXT_CHARACTERS,
        allowEmpty = true,
    )
    val duration = raw["duration_ms"]?.let { integerValue(it, "transcription.duration_ms") }
    val processing = raw["processing_ms"]?.let { integerValue(it, "transcription.processing_ms") }

    val segmentValues = raw["segments"] ?: JsonArray(emptyList())
    val segmentArray = segmentValues as? JsonArray ?: fail("transcription.segments must be an array.")
    if (segmentArray.size > 10_000) fail("transcription.segments is too large.")
    val segments = segmentArray.mapIndexed { index, item ->
        val path = "transcription.segments[$index]"
        val segment = objectValue(item, path)
        val start = integerValue(required(segment, "start_ms", path), "$path.start_ms")
        val end = integerValue(required(segment, "end_ms", path), "$path.end_ms")
        if (end < start) fail("$path.end_ms must not precede start_ms.")
        if (duration != null && end > duration) fail("$path.end_ms exceeds transcription.duration_ms.")
        TranscriptionSegmentV1(
            startMs = start,
            endMs = end,
            text = stringValue(required(segment, "text", path), "$path.text", MAX_TEXT_CHARACTERS, true),
        )
    }

    val warningValues = raw["warnings"] ?: JsonArray(emptyList())
    val warningArray = warningValues as? JsonArray ?: fail("transcription.warnings must be an array.")
    if (warningArray.size > 64) fail("transcription.warnings is too large.")
    val warnings = warningArray.mapIndexed { index, item ->
        stringValue(item, "transcription.warnings[$index]", 1_000, true)
    }

    val error = if (status == "failed") {
        if (text.isNotEmpty()) fail("failed transcription text must be empty.")
        val errorValue = objectValue(required(raw, "error", "transcription"), "transcription.error")
        val retryable = (required(errorValue, "retryable", "transcription.error") as? JsonPrimitive)
            ?.booleanOrNull ?: fail("transcription.error.retryable must be a boolean.")
        TranscriptionErrorV1(
            code = stringValue(required(errorValue, "code", "transcription.error"), "transcription.error.code", 128),
            message = stringValue(
                required(errorValue, "message", "transcription.error"),
                "transcription.error.message",
                4_000,
            ),
            retryable = retryable,
        )
    } else {
        if (raw["error"] != null) fail("only failed transcriptions may contain error.")
        if (status in setOf("no_speech", "cancelled") && text.isNotEmpty()) {
            fail("$status transcription text must be empty.")
        }
        null
    }

    return TranscriptionResultV1(
        requestId = requestId,
        status = status,
        text = text,
        provider = optionalString(raw, "provider", "transcription", 128),
        language = optionalString(raw, "language", "transcription", 64),
        durationMs = duration,
        processingMs = processing,
        model = optionalString(raw, "model", "transcription", 128),
        segments = segments,
        warnings = warnings,
        error = error,
    )
}

private inline fun <T> parseContract(block: () -> T): VoiceContractParseResult<T> =
    try {
        VoiceContractParseResult.Verified(block())
    } catch (error: VoiceContractIncompatibleException) {
        VoiceContractParseResult.Incompatible(error.contract, error.version)
    } catch (error: VoiceContractDecodeException) {
        VoiceContractParseResult.Invalid(error.message ?: "Voice contract is malformed.")
    } catch (_: Exception) {
        VoiceContractParseResult.Invalid("Voice contract is malformed.")
    }

fun parseTranscriptionResult(value: JsonElement): VoiceContractParseResult<TranscriptionResultV1> =
    parseContract { parseTranscription(value) }

fun parsePhoneAudio(value: JsonElement): VoiceContractParseResult<PhoneAudioEnvelopeV1> =
    parseContract {
        val raw = objectValue(value, "phone_audio")
        val contract = stringValue(required(raw, "contract", "phone_audio"), "phone_audio.contract", 128)
        if (contract != "fabric.phone_audio") fail("phone_audio.contract is unsupported.")
        val version = (required(raw, "version", "phone_audio") as? JsonPrimitive)?.intOrNull
            ?: fail("phone_audio.version must be an integer.")
        if (version != FABRIC_PHONE_AUDIO_CONTRACT_VERSION) {
            throw VoiceContractIncompatibleException(contract, version)
        }
        val mode = stringValue(required(raw, "mode", "phone_audio"), "phone_audio.mode", 32)
        if (mode !in PHONE_AUDIO_MODES) fail("phone_audio.mode is invalid.")
        val mimeType = stringValue(
            required(raw, "mime_type", "phone_audio"),
            "phone_audio.mime_type",
            128,
        ).lowercase()
        if (!mimeType.startsWith("audio/") && mimeType != "video/webm") {
            fail("phone_audio.mime_type must describe audio.")
        }
        PhoneAudioEnvelopeV1(
            captureId = stringValue(
                required(raw, "capture_id", "phone_audio"),
                "phone_audio.capture_id",
                128,
            ),
            mode = mode,
            mimeType = mimeType,
            durationMs = integerValue(required(raw, "duration_ms", "phone_audio"), "phone_audio.duration_ms"),
            result = parseTranscription(required(raw, "result", "phone_audio")),
        )
    }
